### DISCUSSION TITLE
Best practice for despawning entities

### DISCUSSION THREAD

**Discussion by shouya** [bevyengine/bevy#2658] [Q&A] [answered]
I am new to game developing, especially ECS. I wonder what's the best practice of despawning objects.

Say I want to make a simple space shooter, there are two ways an entities should be despawned:

1. when a bullet or an enemy aircraft move out of bound
2. when an enemy's HP reaches zero

So basically, I have two systems:

1. `object_movement` system, which iterate through all eligible objects (incl. bullet and enemy aircrafts) and perform `obj.position += obj.velocity`, if they are out of bound, despawn them.
2. `enemy_hp_deduction` system, it handles player bullet collision events with enemies and deduct their hp. when their hp reaches zero, despawn them.

These two systems, IMHO, should be independent. But they both need to despwan the objects. However, if `object_movement` system runs first and despawned an enemy aircraft using `Commands.entity(enemy_aircraft).despawn()`, and soon follows, the `enemy_hp_deduction` system may fail because it may try to modify the `hp` values for the enemy aircraft. Similar thing can happen if `enemy_hp_deduction` runs prior to `object_movement`.

How I currently deal with it is to add a `ShouldDespawn` tag component to the objects instead of actually despawning them. Then on the `PostUpdate` stage I use a dedicated system to despwan every entities that have this tag. But I still find it awkward because ideally, I hope that an entity despawned shouldn't be picked up by later systems. For example, if an enemy aircraft is out of bound and despawned, I don't even have to consider modifing its HP values.

I searched the internet but couldn't find a satisfactory solution to it. So I'm looking to be imparted some insights from you veteran game developers :)

**Comment by afonsolage** [comment, marked as answer]
That's a good question, I'm also new, but I can think in a simple, but maybe naive solution:

1- `out_of_bounds` system, where you check if the object is out of bounds and despawn it;
2- `remove_dead` system, where you just check if hp is zero and remove enemies;

In other words, move the despawning to it's own system, based on the shared data between those systems.

**Reply by shouya** [reply]
Neat! That I have not thought of.

There is one question though - if an entity happens to be just moved out of bound and killed at the same time (hp becomes zero), the entity will be removed twice. I wonder if bevy would crash by double despawning an entity?

**Reply by shouya** [reply]
I did a little experiment, and it turns out despawning twice is totally valid! It seems that only modifying a component of a despawned entity will cause it to crash. I think I'm going to go with this solution, thank you @afonsolage!

**Reply by afonsolage** [reply]
if I'm not mistaken, it wont crash, instead a debug error will be printed: `Failed to despawn non-existent entity`. Since all commands are run at the end of the stage, if you have issues, you can have separated stages, like one stage for moving things and another stage for checking combat things (like physical stage and combat stage).

**Comment by Rexcrazy804** [comment]
This is very late, but I'd like to suggest an alternative. You can first create a `struct DespawnEvent(Entity)` Event and write into that from any of the systems that could check whether an entity has gone beyond bounds or has reach 0 hp. After passing the entity using the event you can then use an event reader to iterate through and despawn the entities. You can optimize this further by adding the` .run_if(on_event::<DespawnEvent>())` to the system that actually despawns the entity to ensure that it is only executed if an entity has died. Hope this helped (or rather I am a 'little' too late, aren't I?)

**Reply by Qwyntex** [reply]
This has the advantage of being easily implemented for all kinds of Entities, not just ones that have a specific component (like HP), allowing for a centralized handling of despawning. With this you also could for example also give a reason or cause for the despawning and have all that neatly handled by a single system which can be easily adapted if necessary.

There might still be a better solution somewhere but I find this one works the best for me and my current project and would recommend using this. (But you may have to take this with a grain of salt, I haven't been using bevy for that long)

**Reply by Rexcrazy804** [reply]
haha, that is quite neat indeed! the other advantage I noticed is that the function signature will have the DespawnEventWritter, making it easier to identify functions that despawn entities.
